pkg/driver/node.go

```diff
@@ -0,0 +1,609 @@
+			logrus.Errorf("failed to mkdir %s, error", mntPath)
+			return err
+		}
+	}
+
+	if !notMnt {
+		logrus.Infof("Volume %s has been mounted already at %v", req.GetVolumeId(), mntPath)
+		return nil
+	}
+
+	fsType := req.GetVolumeCapability().GetMount().GetFsType()
+	options := []string{}
+	mountFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+	options = append(options, mountFlags...)
+
+	err = ns.mounter.FormatAndMount(devicePath, mntPath, fsType, options)
+	if err != nil {
+		logrus.Errorf(
+			"Failed to mount iscsi volume %s [%s, %s] to %s, error %v",
+			req.GetVolumeId(), devicePath, fsType, mntPath, err,
+		)
+		return err
+	}
+	return nil
+}
+
+// NodePublishVolume publishes (mounts) the volume
+// at the corresponding node at a given path
+//
+// This implements csi.NodeServer
+func (ns *node) NodePublishVolume(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest,
+) (*csi.NodePublishVolumeResponse, error) {
+
+	volumeID := req.GetVolumeId()
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.FailedPrecondition, "Volume ID not provided")
+	}
+
+	target := req.GetTargetPath()
+	if len(target) == 0 {
+		return nil, status.Error(codes.FailedPrecondition, "Target path not provided")
+	}
+
+	volCap := req.GetVolumeCapability()
+	if volCap == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Volume capability not provided")
+	}
+
+	if !isValidVolumeCapabilities([]*csi.VolumeCapability{volCap}) {
+		return nil, status.Error(codes.InvalidArgument, "Volume capability not supported")
+	}
+
+	mountOptions := []string{"bind"}
+	if req.GetReadonly() {
+		mountOptions = append(mountOptions, "ro")
+	}
+	switch mode := volCap.GetAccessType().(type) {
+	case *csi.VolumeCapability_Block:
+		return &csi.NodePublishVolumeResponse{}, status.Error(codes.Unimplemented, "doesn't support block device provisioning")
+	case *csi.VolumeCapability_Mount:
+		if err := ns.nodePublishVolumeForFileSystem(req, mountOptions, mode); err != nil {
+			return nil, err
+		}
+	}
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func (ns *node) nodePublishVolumeForFileSystem(req *csi.NodePublishVolumeRequest, mountOptions []string, mode *csi.VolumeCapability_Mount) error {
+	target := req.GetTargetPath()
+	source := req.GetStagingTargetPath()
+	if m := mode.Mount; m != nil {
+		hasOption := func(options []string, opt string) bool {
+			for _, o := range options {
+				if o == opt {
+					return true
+				}
+			}
+			return false
+		}
+		for _, f := range m.MountFlags {
+			if !hasOption(mountOptions, f) {
+				mountOptions = append(mountOptions, f)
+			}
+		}
+	}
+
+	logrus.Infof("NodePublishVolume: creating dir %s", target)
+	if err := os.MkdirAll(target, 0000); err != nil {
+		return status.Errorf(codes.Internal, "Could not create dir {%q}, err: %v", target, err)
+	}
+
+	// in case if the dir already exists, above call returns nil
+	// so permission needs to be updated
+	if err := os.Chmod(target, 0000); err != nil {
```

Comment:
why 0000 permission.

Reply:
This is done to avoid reading and writing on the path by the app when it is not mounted.

Reply:
please update the comment mentinong the above @utkarshmani1997 .